File: python/stack/sequence_stack.py

class ArrayStack(object):
    '''
    data : data array for stack
    size : data size this stack can store
    count : current data number in stack
    '''
    def __init__(self, num):
        self.data = list(None for _ in range(1, num))
        self.size = num
        self.count = 0

    def push(self, data):
        '''
        on success, return True; or return False
        '''
        
        if self.count == self.size:
            print('stack is full, push fail')
            return False

        self.count += 1
        self.data[self.count] = data

        print('push element')

        return True

    def pop(self):
        '''
        on success, return value; or return None
        '''
        if self.count == 0:
            return None

        data = self.data[self.count] 
        self.count -= 1

        return data


def main():
    stack = ArrayStack(5)

    stack.push(1)
    stack.push('abc')

    print('data1 : %s' % stack.pop())
    print('data2 : %s' % stack.pop())

    stack.push(4)
    print('data2 : %s' % stack.pop())

if __name__ == '__main__':
    main()

File: c/stack/sequence_stack.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

struct array_stack {
    unsigned int *data;
    int count;
    int size;
};

bool init_stack(struct array_stack* stack, int size) {
    stack->data = (int *)malloc(sizeof(int) * size);
    if (stack->data == NULL)
        return false;
    else {
        stack->count = 0;
        stack->size = size;
    }

    return true;
}

bool push(struct array_stack* stack, int data) {

    if (stack->count > stack->size) {
        printf("stack if full, push fail\n");
        return false;
    }

    stack->data[stack->count] = data;
    stack->count++;

    return true;
}

int pop(struct array_stack* stack) {
    struct array_stack node;

    stack->count--;
    if (stack->count < 0) {
        printf("stack is empty, pop fail\n");
        stack->count = 0;
        return -1;
    }

    return stack->data[stack->count];

}

int main() {
    struct array_stack stack;
    bool ret = false;

    ret = init_stack(&stack, 10);
    printf("stack init success\n");

    push(&stack, 1);
    push(&stack, 2);

    printf("data : %d\n", pop(&stack));
    printf("data : %d\n", pop(&stack));
    printf("data : %d\n", pop(&stack));

    push(&stack, 3);
    printf("data : %d\n", pop(&stack));

    return 1;
}